pyvista/core/filters/data_set.py

```diff
@@ -2004,23 +1996,43 @@ def glyph(
                 )
                 orient = False
 
+        set_active_scalars_vectors = False
+
         if scale and orient:
-            if (
-                dataset.active_vectors_info.association == FieldAssociation.CELL
-                and dataset.active_scalars_info.association == FieldAssociation.CELL
-            ):
-                source_data = dataset.cell_centers()
-            elif (
-                dataset.active_vectors_info.association == FieldAssociation.POINT
-                and dataset.active_scalars_info.association == FieldAssociation.POINT
-            ):
-                source_data = dataset
-            else:
-                raise ValueError(
-                    "Both ``scale`` and ``orient`` must use " "point data or cell data."
-                )
-        else:
-            source_data = dataset
+            if dataset.active_vectors_info.association != dataset.active_scalars_info.association:
+                raise ValueError("Both ``scale`` and ``orient`` must use point data or cell data.")
+
+        source_data = dataset
+        if (scale and dataset.active_scalars_info.association == FieldAssociation.CELL) or (
+            orient and dataset.active_vectors_info.association == FieldAssociation.CELL
+        ):
+            source_data = dataset.cell_centers()
+            set_active_scalars_vectors = True
+
+        # Clean the points before glyphing
+        if tolerance is not None:
+            small = pyvista.PolyData(source_data.points)
+            small.point_data.update(source_data.point_data)
+            source_data = small.clean(
+                point_merging=True,
+                merge_tol=tolerance,
+                lines_to_points=False,
+                polys_to_lines=False,
+                strips_to_polys=False,
+                inplace=False,
+                absolute=absolute,
+                progress_bar=progress_bar,
+            )
+            set_active_scalars_vectors = True
+
+        # converting from cell -> points and the point merging operation both destroy the active vectors, so set them
+        # again
+        if set_active_scalars_vectors:
+            if scale:
+                source_data.set_active_scalars(dataset.active_scalars_name, 'point')
```

Comment:
If we revert preference above, this would need to be changed here.

Reply:
At this point, `source_data` only consists of points.
- In the case of point data, `source_data` is just the original `dataset`
- In the case of cell data, `source_data = dataset.cell_centers()`
So point should always be the "preference," here, but it could also be strictly enforced if we had the methods to do so.

Reply:
You are right.  this is good.